app/services/toolbox/dalle.rb

```diff
@@ -0,0 +1,19 @@
+class Toolbox::Dalle < Toolbox
+
+  describe :generate_an_image, <<~S
+    Generate an image based on what the user asks you to generate. You will pass the user's prompt and will get back a URL to an image.
```

Comment:
I think it's that simple! I wonder how long those URLs from OpenAI stay active? And I wonder how the GPT currently handles displaying that URL. Probably it'll just format it into markdown and it'll be clickable, which is a nice default.

Reply:
@krschacht according to the documentation the images are available for roughly one hour. I hacked together a little download (see code), but I don't understand how I can control the output to the message to have the local URL rather than the one that I get from the dall-e API. Looks like you are doing it the right way with the weather API, but I don't quite understand how it works. Maybe you can help out.

Reply:
Oh, just use active storage. Although I say that and I’m not actually super familiar with active storage. I’ve pretty much always used it by following basic tutorials for file uploads. But it should be the standard “active storage way.”

Basically, Every Message has_many Document. Those are basically “attachments” (i couldn’t call it File b/c that’s a reserved word in ruby). But a Document is one-to-one a File. It has an active storage attachment on it.

I believe that once you have the URL you can do `message.documents.new` and `document.file.attach(…)` I think this is the reference:
https://guides.rubyonrails.org/active_storage_overview.html#attaching-file-io-objects

I think then simply `document.save!` and the file will be stored on our servers. Notably, our active storage backend is postgres so it’ll be serialized and stored in the DB. I just did that to avoid setting up S3 and avoid the missing persistent-disk storage on so many hosts (like heroku).

I think once the image is saved as a Document attached to a Message, it should kick on all the display code that’s within the _message partial. Code is already in place to display a fallback spinner until the image’s various sizes are processed. And it displays a small image as a thumbnail and a bigger one when you click on it. To trigger all this code just try attaching an image right now in a conversation thread and you’ll see all that work. I don’t think we’ll need to modify any of that, once you get the image attached.

Reply:
Yeah, ActiveStorage is the way to go for sure.
But I mean, the message has to be modified once saved so that the URL to the image can be extracted, saved via ActiveStorage and then the markup needs to be replaced with the URL of the image from ActiveStorage. I guess this could be a background task that looks at a new message that is being created if it has an image URL inside of it or not... hmmmm, do you understand what I mean? Do you agree with the approach?

Reply:
I think you might be thinking it's more complicated than it is.

There are 4 messages involved. Let me see if I can summarize all this context accurately off the top of my head:

A user message:

1. Message.user which is someone saying "make me an image of a cat with a birthday hat"
2. Message.content_tool_calls  which is the API responding with a request to call a tool. Right as this message is being saved by the normal get_next_mesage_job it, only the fly, creates an additional message
3. Message.tool, the requested tool from the content_tool_calls is actually called and it's response is saved
4. Message.assistant is our system requesting yet again for the assistant to generate a response, after reviewing the full conversation history which now includes these tool calls

The (3) and (4) happen really quickly in the "call tools before wrapping up" method near the bottom of get_next_message_job.

So... it's on (3) when your tool will be called. The tool returns a JSON response which will include the URL. Where to handle that? One of these messages should just get a *.document attached to it, but I'm not sure which one. I think the most logical one is (4) since it's going to include some friendly text that the API generates such as "Here is your cat". And if that message has an attached document, then we don't have to think at all about markup replacing the URL of an image. `_message` partial automatically renders an image if one is attached to a message as a document.

And we don't need to think about doing any processing in a background job, because all of this is happening in a background job.

Ohhh, here is where we do it:
https://github.com/AllYourBot/hostedgpt/blob/b56f1de2d347ad7a4ee191ce36d658c067e6b7d3/app/jobs/get_next_ai_message_job.rb#L177

I forgot, message (4) is created within the initial get_next_message_job but it's created empty and then this job ends by kicking off a new job. So I think right on the line I linked to: where we are stubbing out the final assistants response, attach the document right there. But the logic is a little complicated to know.

I think I'd suggest this:

* Turn the L177 create! block into a new() block
* Then call a helper method such as `after_tool_processing(assistant_reply, tool_messages)` (it's called `msgs` above but we should rename that `tool_messages` for clarity)
* Inside this helper, you loop through the tool_messages and examine the contents. If it's the particular tool you created then you do some quick modifications to `assistant_reply` with `.documents.new()`

Does that make sense? As I reflect, this is a bit of a naive implementation. Ideally, a Toolbox child class should be able to contain an `after_run` hook which is given the message as an input and can do whatever it wants with it. That would be a much cleaner design. But I think it feels like a premature abstraction. I want to run into at least a couple more situations where tool responses need to modify the assistant's message before trying to generalize it.

After writing all this I started off thinking "this isn't so complicated..." and now I realize it is complicated :) but hopefully this proposed implementation guides the way. I don't actually think it's a lot of code to write, it's really just groking the mental model that's the hard part.

Reply:
Thanks for the in-depth explanation of the flow. I will take a look

Reply:
Ok, I understand what you mean. Here is the ```Message``` that I got as a reply:

```
=> #<Message:0x0000ffff9b0cc6c8
 id: 1048909790,
 conversation_id: 962582977,
 role: "assistant",
 content_text:
  "Here is an image of a bottle:\n\n![Bottle](https://oaidalleapiprodscus.blob.core.windows.net/private/org-3gmrqp9cIUIbQLYMoIkhGTAb/user-l5OLBM3CmF42GH5rlcxG3eQH/img-P0sNHzxaDyBXtRsqmzI5g1TS.png?st=2024-06-13T11%3A58%3A37Z&se=2024-06-13T13%3A58%3A37Z&sp=r&sv=2023-11-03&sr=b&rscd=inline&rsct=image/png&skoid=6aaadede-4fb3-4698-a8f6-684d7786b067&sktid=a48cca56-e6da-484e-a814-9c849652bcb3&skt=2024-06-12T18%3A49%3A07Z&ske=2024-06-13T18%3A49%3A07Z&sks=b&skv=2023-11-03&sig=4nxEJvSJmaIK%2BNnrrn1c%2BiLd3JSkJudSvYhSHiIv9uk%3D)",
 created_at: Thu, 13 Jun 2024 12:58:49.176219000 UTC +00:00,
 updated_at: Thu, 13 Jun 2024 12:58:56.704132000 UTC +00:00,
 content_document_id: nil,
 run_id: nil,
 assistant_id: 1013349885,
 processed_at: Thu, 13 Jun 2024 12:58:49.259744000 UTC +00:00,
 index: 3,
 version: 2,
 cancelled_at: nil,
 branched: false,
 branched_from_version: nil,
 content_tool_calls: {},
 tool_call_id: nil,
 versions: "">
```


As you can see, the `.content_text` contains the Markdown including an image link. I think this part needs to be extracted, image downloaded and attached to the Message object like you described and then this part of the ```.content_text``` markup needs to be removed.

Is it like this, or am I missing something?